src/components/utils/include/utils/timer.h

```diff
@@ -84,47 +90,127 @@ class Timer {
-   * and if we puts to timer_create zero timeout then we get sys error(22)
+   * @brief GetTimeout
+   * @return returns timeout
    */
-  void SetTimeoutUnsafe(const Milliseconds timeout);
+  Milliseconds GetTimeout() const;
 
+ private:
   /**
-   * @brief startUnsafe, stopUnsafe - methods used for correct synchronization
-   *  and must be used only with sync_primitive (auto_lock e.g.)
-   * @return true if start/stop successfull, false if unsuccessfull
+   * @brief Delegate release timer, will call callback function one time
+   * or call delegate every timeout function while stop()
+   * won't be called. It's depend on flag.
    */
-  void StartUnsafe();
-  bool StopUnsafe();
+  class TimerDelegate : public threads::ThreadDelegate {
+   public:
+    /**
+     * @brief Default constructor
+     *
+     * @param timer_thread The Timer_thread pointer
+     */
+    explicit TimerDelegate(Timer* timer);
+
+    /**
+     * @brief Destructor
+     */
+    virtual ~TimerDelegate();
+
+    /**
+     * @brief Thread main function.
+     */
+    void threadMain() OVERRIDE;
+
+    /**
+     * @brief Called by thread::thread to free all allocated resources.
+     */
+    void exitThreadMain() OVERRIDE;
+
+    /**
+     * @brief Set new Timeout
+     * @param timeout_milliseconds New timeout to be set
+     */
+    virtual void SetTimeOut(const uint32_t timeout_milliseconds);
+
+    /**
+      * @brief Wait until timer will start
+      */
+    virtual void WaitUntilStart();
+
+    /**
+     * @brief Quits threadMain function after next loop.
+     */
+    virtual void ShouldBeStopped();
+
+    /**
+     * @brief Restarts non-loop timer after current iteration.
+     */
+    virtual void ShouldBeRestarted();
+
+    /**
+     * @brief Return flag IsGoingStop
+     */
+    bool IsGoingToStop() const;
+
+    /**
+     * @brief Gets timeout with overflow check
+     * @return timeout
+     */
+    inline int32_t get_timeout() const {
+      return std::min(
+          static_cast<uint32_t>(std::numeric_limits<int32_t>::max()),
+          timeout_milliseconds_);
+    }
+
+    /**
+      * @brief Make class delegate repeatable
+      */
+    inline void make_repeatable() {
+      is_repeatable_ = true;
+    }
+
+   protected:
+    Timer* timer_;
+    uint32_t timeout_milliseconds_;
+    sync_primitives::Lock state_lock_;
+    sync_primitives::ConditionalVariable termination_condition_;
+    sync_primitives::ConditionalVariable starting_condition_;
+    sync_primitives::atomic_bool stop_flag_;
+    sync_primitives::atomic_bool restart_flag_;
+    sync_primitives::atomic_bool is_started_flag_;
+    sync_primitives::atomic_bool is_repeatable_;
+
+   private:
+    DISALLOW_COPY_AND_ASSIGN(TimerDelegate);
+  };
+
+  const std::string name_;
+  std::auto_ptr<TimerTask> task_;
+  uint32_t timeout_ms_;
+  sync_primitives::atomic_bool is_running_;
+  TimerDelegate delegate_;
+  threads::Thread* thread_;
+
```

Comment:
@anosach-luxoft , yeah, you right. `is_running` is useless flag. I'll delete it. Other flags have different purpose and can't be minimize.See cc02161